File: src/main/java/io/github/tanguygab/tabadditions/shared/features/advancedconditions/AdvancedConditions.java
package io.github.tanguygab.tabadditions.shared.features.advancedconditions;

import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Function;

import lombok.Getter;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.placeholders.TabPlaceholder;
import me.neznamy.tab.shared.platform.TabPlayer;
import me.neznamy.tab.shared.features.PlaceholderManagerImpl;

/**
 * The main condition class. It allows users to configure different
 * condition types that must be met in order to display specified
 * text or make a condition requirement for a visual to be displayed.
 */
public class AdvancedConditions {

    /** All conditions defined in configuration including anonymous conditions */
    private static Map<String, AdvancedConditions> registeredConditions = new HashMap<>();

    /** All supported sub-condition types */
    @Getter private static final Map<String, Function<String, BiFunction<TabPlayer, TabPlayer, Boolean>>> conditionTypes =
            new LinkedHashMap<>() {{
                put(">=", line -> new NumericCondition(line.split(">="), (left, right) -> left >= right)::isMet);
                put(">", line -> new NumericCondition(line.split(">"), (left, right) -> left > right)::isMet);
                put("<=", line -> new NumericCondition(line.split("<="), (left, right) -> left <= right)::isMet);
                put("<-", line -> new StringCondition(line.split("<-"), String::contains)::isMet);
                put("<", line -> new NumericCondition(line.split("<"), (left, right) -> left < right)::isMet);
                put("|-", line -> new StringCondition(line.split("|-"), String::startsWith)::isMet);
                put("-|", line -> new StringCondition(line.split("-|"), String::endsWith)::isMet);
                put("!=", line -> new StringCondition(line.split("!="), (left, right) -> !left.equals(right))::isMet);
                put("=", line -> new StringCondition(line.split("="), String::equals)::isMet);
                put("permission:", line -> (viewer,target) -> viewer == null || viewer.hasPermission(line.split(":")[1]));
            }};

    /** Name of this condition defined in configuration */
    @Getter private final String name;

    /** All defined sub-conditions inside this conditions */
    protected Map<BiFunction<TabPlayer, TabPlayer, Boolean>,String> subConditions = new LinkedHashMap<>();

    /**
     * Refresh interval of placeholder created from this condition.
     * It is calculated based on nested placeholders used in sub-conditions.
     */
    @Getter private int refresh = -1;

    /** List of all placeholders used inside this condition */
    private final List<String> placeholdersInConditions = new ArrayList<>();
    /** Boolean that returns whether the condition has relational placeholders or not */
    private boolean hasRelationalPlaceholders;

    /**
     * Constructs new instance with given parameters and registers
     * this condition to list as well as the placeholder.
     *
     * @param   name
     *          name of condition
     * @param   conditions
     *          map of condition lines and their output
     */
    public AdvancedConditions(String name, Map<String,String> conditions) {
        this.name = name;
        if (conditions == null) {
            TAB.getInstance().getMisconfigurationHelper().conditionHasNoConditions(name);
            return;
        }
        PlaceholderManagerImpl pm = TAB.getInstance().getPlaceholderManager();
        conditions.forEach((line,text)->{
            BiFunction<TabPlayer, TabPlayer, Boolean> condition = compile(line);
            if (condition == null && !line.equals("else")) {
                TAB.getInstance().getMisconfigurationHelper().invalidConditionPattern("[TAB-Additions]-"+name, line);
                return;
            }
            subConditions.put(condition,text);
            if (!line.startsWith("permission:")) placeholdersInConditions.addAll(pm.detectPlaceholders(line));
            else if (refresh > 1000 || refresh == -1) refresh = 1000; //permission refreshing will be done every second
            placeholdersInConditions.addAll(pm.detectPlaceholders(text));
        });
        hasRelationalPlaceholders = placeholdersInConditions.stream().anyMatch(placeholder->placeholder.startsWith("%rel_"));
        registeredConditions.put(name, this);

        if (hasRelationalPlaceholders) pm.registerRelationalPlaceholder("%rel_condition+:"+name+"%", refresh, (viewer,target) -> getText((TabPlayer) viewer,(TabPlayer) target));
        else pm.registerPlayerPlaceholder("%condition+:"+name+"%", refresh, p -> getText((TabPlayer) p,null));
    }

    /**
     * Configures refresh interval and registers nested placeholders
     */
    public void finishSetup() {
        PlaceholderManagerImpl pm = TAB.getInstance().getPlaceholderManager();
        for (String placeholder : placeholdersInConditions) {
            TabPlaceholder pl = pm.getPlaceholder(placeholder);
            pl.addParent((hasRelationalPlaceholders ? "%rel_" : "%") +"condition+:"+name+"%");
            if (pl.getRefresh() < refresh && pl.getRefresh() != -1) refresh = pl.getRefresh();
        }
        pm.addUsedPlaceholders(placeholdersInConditions);
    }

    /**
     * Returns text for player based on if condition is met or not
     *
     * @return  yes or no value depending on if condition passed or not
     */
    public String getText(TabPlayer viewer, TabPlayer target) {
        return subConditions.get(check(viewer,target));
    }

    /**
     * Returns {@code true} if condition is met for player, {@code false} if not
     *
     * @return  {@code true} if met, {@code false} if not
     */
    public boolean isMet(TabPlayer viewer, TabPlayer target) {
        return check(viewer,target) != null;
    }

    /**
     * Returns {@code true} if condition is met for player, {@code false} if not
     *
     * @return  {@code true} if met, {@code false} if not
     */
    public boolean isMet(TabPlayer viewer) {
        return check(viewer,viewer) != null;
    }

    public BiFunction<TabPlayer,TabPlayer,Boolean> check(TabPlayer viewer, TabPlayer target) {
        for (BiFunction<TabPlayer, TabPlayer, Boolean> condition : subConditions.keySet())
            if (condition == null || condition.apply(viewer,target)) return condition;
        return null;
    }

    /**
     * Returns condition from given string. If the string is name of a condition,
     * that condition is returned. If it's a condition pattern, it is compiled and
     * returned. If the string is null, null is returned.
     *
     * @param   string
     *          condition name or pattern
     * @return  condition from string
     */
    public static AdvancedConditions getCondition(String string) {
        if (string == null || string.equals("")) return null;
        if (registeredConditions.containsKey(string))
            return registeredConditions.get(string);

        Map<String,String> conditions = new HashMap<>();
        conditions.put(string,"");
        AdvancedConditions c = new AdvancedConditions("AnonymousCondition["+string+"]", conditions);
        c.finishSetup();
        return c;
    }

    /**
     * Clears registered condition map on plugin reload
     */
    public static void clearConditions() {
        registeredConditions = new HashMap<>();
    }

    /**
     * Marks all placeholders used in the condition as used and registers them.
     * Using a separate method to avoid premature registration of nested conditional placeholders
     * before they are registered properly.
     */
    public static void finishSetups() {
        registeredConditions.values().forEach(AdvancedConditions::finishSetup);
    }

    /**
     * Compiles condition from condition line. This includes detection
     * what kind of condition it is and creating it.
     *
     * @param   line
     *          condition line
     * @return  compiled condition or null if no valid pattern was found
     */
    private static BiFunction<TabPlayer, TabPlayer, Boolean> compile(String line) {
        for (Map.Entry<String, Function<String, BiFunction<TabPlayer, TabPlayer, Boolean>>> entry : AdvancedConditions.getConditionTypes().entrySet())
            if (line.contains(entry.getKey())) return entry.getValue().apply(line);
        return null;
    }
}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/advancedconditions/AdvancedConditions.kt
package io.github.tanguygab.tabadditions.shared.features.advancedconditions

import me.neznamy.tab.shared.TAB
import me.neznamy.tab.shared.features.PlaceholderManagerImpl
import me.neznamy.tab.shared.platform.TabPlayer

/**
 * The main condition class. It allows users to configure different
 * condition types that must be met in order to display specified
 * text or make a condition requirement for a visual to be displayed.
 */
class AdvancedConditions(
    /** Name of this condition defined in configuration  */
    val name: String, conditions: Map<String, String>
) {
    /** All defined sub-conditions inside this conditions  */
    private var subConditions = mutableMapOf<((TabPlayer, TabPlayer) -> Boolean)?, String>()

    /**
     * Refresh interval of placeholder created from this condition.
     * It is calculated based on nested placeholders used in sub-conditions.
     */
    var refresh = -1

    /** List of all placeholders used inside this condition  */
    private val placeholdersInConditions = mutableListOf<String>()

    /** Boolean that returns whether the condition has relational placeholders or not  */
    private var hasRelationalPlaceholders: Boolean

    /**
     * Constructs new instance with given parameters and registers
     * this condition to list as well as the placeholder.
     *
     * @param   name
     * name of condition
     * @param   conditions
     * map of condition lines and their output
     */
    init {
        conditions.forEach { (line: String, text: String) ->
            val condition = compile(line)
            if (condition == null && line != "else") return@forEach

            subConditions[condition] = text
            if (!line.startsWith("permission:")) placeholdersInConditions.addAll(
                PlaceholderManagerImpl.detectPlaceholders(line)
            ) else if (refresh > 1000 || refresh == -1) refresh = 1000 //permission refreshing will be done every second
            placeholdersInConditions.addAll(PlaceholderManagerImpl.detectPlaceholders(text))
        }
        hasRelationalPlaceholders = placeholdersInConditions.any { it.startsWith("%rel_") }
        registeredConditions[name] = this

        val pm = TAB.getInstance().placeholderManager
        if (hasRelationalPlaceholders) pm.registerRelationalPlaceholder("%rel_condition+:$name%", refresh)
        { viewer, target -> getText(viewer as TabPlayer, target as TabPlayer) }
        else pm.registerPlayerPlaceholder("%condition+:$name%", refresh)
        { getText(it as TabPlayer, it) }
    }

    /**
     * Configures refresh interval and registers nested placeholders
     */
    fun finishSetup() {
        val pm = TAB.getInstance().placeholderManager
        for (placeholder in placeholdersInConditions) {
            val pl = pm.getPlaceholder(placeholder)
            pl.addParent((if (hasRelationalPlaceholders) "%rel_" else "%") + "condition+:" + name + "%")
            if (pl.refresh < refresh && pl.refresh != -1) refresh = pl.refresh
        }
        pm.addUsedPlaceholders(placeholdersInConditions)
    }

    /**
     * Returns text for player based on if condition is met or not
     *
     * @return  yes or no value depending on if condition passed or not
     */
    fun getText(viewer: TabPlayer, target: TabPlayer) = subConditions[check(viewer, target)]

    /**
     * Returns `true` if condition is met for player, `false` if not
     *
     * @return  `true` if met, `false` if not
     */
    fun isMet(viewer: TabPlayer, target: TabPlayer) = check(viewer, target) != null

    /**
     * Returns `true` if condition is met for player, `false` if not
     *
     * @return  `true` if met, `false` if not
     */
    fun isMet(viewer: TabPlayer) = check(viewer, viewer) != null

    fun check(viewer: TabPlayer, target: TabPlayer) = subConditions.keys.find { it == null || it(viewer, target) }

    companion object {
        /** All conditions defined in configuration including anonymous conditions  */
        private val registeredConditions = mutableMapOf<String, AdvancedConditions>()

        /** All supported sub-condition types  */
        private val conditionTypes = mutableMapOf<String, (String) -> (TabPlayer, TabPlayer) -> Boolean>(
            ">=" to { { viewer: TabPlayer, target: TabPlayer ->
                NumericCondition(it.split(">="))
                { left: Double, right: Double -> left >= right }.isMet(viewer, target)
            } },
            ">" to { { viewer: TabPlayer, target: TabPlayer ->
                NumericCondition(it.split(">"))
                { left: Double, right: Double -> left > right }.isMet(viewer, target)
            } },
            "<=" to { { viewer: TabPlayer, target: TabPlayer ->
                NumericCondition(it.split("<="))
                { left: Double, right: Double -> left <= right }.isMet(viewer, target)
            } },
            "<-" to { { viewer: TabPlayer, target: TabPlayer ->
                StringCondition(it.split("<-"))
                { obj: String, s: String -> s in obj }.isMet(viewer, target)
            } },
            "<" to { { viewer: TabPlayer, target: TabPlayer ->
                NumericCondition(it.split("<"))
                { left: Double, right: Double -> left < right }.isMet(viewer, target)
            } },
            "|-" to { { viewer: TabPlayer, target: TabPlayer ->
                StringCondition(it.split("|-"))
                { obj: String, prefix: String -> obj.startsWith(prefix) }.isMet(viewer, target)
            } },
            "-|" to { { viewer: TabPlayer, target: TabPlayer ->
                StringCondition(it.split("-|"))
                { obj: String, suffix: String -> obj.endsWith(suffix) }.isMet(viewer, target)
            } }
            ,
            "!=" to { { viewer: TabPlayer, target: TabPlayer ->
                StringCondition(it.split("!="))
                { left: String, right: String -> left != right }.isMet(viewer, target)
            } }
            ,
            "=" to { { viewer: TabPlayer, target: TabPlayer ->
                StringCondition(it.split("="))
                { str1: String, str2: String -> str1 == str2 }.isMet(viewer, target)
            } },
            "permission:" to { { viewer: TabPlayer, _: TabPlayer ->
                viewer.hasPermission(it.split(":", limit = 2)[1])
            } }
        )

        /**
         * Returns condition from given string. If the string is name of a condition,
         * that condition is returned. If it's a condition pattern, it is compiled and
         * returned. If the string is null, null is returned.
         *
         * @param   string
         * condition name or pattern
         * @return  condition from string
         */
        fun getCondition(string: String?): AdvancedConditions? {
            if (string.isNullOrEmpty()) return null
            if (string in registeredConditions) return registeredConditions[string]

            val c = AdvancedConditions("AnonymousCondition[$string]", mapOf(string to ""))
            c.finishSetup()
            return c
        }

        /**
         * Clears registered condition map on plugin reload
         */
        fun clearConditions() {
            registeredConditions.clear()
        }

        /**
         * Marks all placeholders used in the condition as used and registers them.
         * Using a separate method to avoid premature registration of nested conditional placeholders
         * before they are registered properly.
         */
        fun finishSetups() {
            registeredConditions.values.forEach { it.finishSetup() }
        }

        /**
         * Compiles condition from condition line. This includes detection
         * what kind of condition it is and creating it.
         *
         * @param   line
         * condition line
         * @return  compiled condition or null if no valid pattern was found
         */
        private fun compile(line: String) = conditionTypes.entries.find { it.key in line }?.value(line)
    }
}